compiler/rustc_session/src/errors.rs

```diff
@@ -324,7 +341,8 @@ pub fn report_lit_error(sess: &ParseSess, err: LitError, lit: token::Lit, span:
             if looks_like_width_suffix(&['i', 'u'], &suf) {
                 // If it looks like a width, try to be helpful.
                 sess.emit_err(InvalidIntLiteralWidth { span, width: suf[1..].into() });
-            } else if let Some(fixed) = fix_base_capitalisation(suf) {
+            } else if let Some(fixed) = fix_base_capitalisation(suf)
+                && looks_like_base_prefixed_number(symbol.as_str(), fix_base_capitalisation(suf)) {
```

Comment:
The repeated call to `fix_base_capitalisation` is a bit clumsy. I think you can change `looks_like_base_prefix_number` to look for an upper-case `X`/`O`/`B`, and then call just it in the condition, and then call `fix_base_capitalisation` in the body?